xterm-color.el

```diff
@@ -497,11 +532,18 @@ in LIFO order."
                                     xterm-color--current-bg
                                     (/= xterm-color--attributes 0)))
           (has? (attrib)       `(/= (logand ,attrib xterm-color--attributes) 0))
-          (face-cache-get ()   `(gethash (logior (ash xterm-color--attributes 16)
-                                                 (ash (or xterm-color--current-bg 0) 8)
-                                                 (or xterm-color--current-fg 0))
+          (face-cache-get ()   `(gethash (let ((fg (or xterm-color--current-fg 0))
+                                               (bg (or xterm-color--current-bg 0)))
+                                           (if (has? +24-bit+)
+                                               (logior (ash xterm-color--attributes 48)
+                                                       (ash bg 24)
+                                                       fg)
```

Comment:
This is broken on 32bit Emacs:

```
ELISP> (ash 1 48)
0 (#o0, #x0, ?\C-@)
```
To fix this, you would use a different scheme to store the color/attribute combination in the hash table. Now, the reason I added the cache to begin with was to avoid consing, since this is the hot path. This implies that a simple scheme of (cons attributes (cons fg bg)) is not ideal. 

I would store fg and bg (48 bits) in a float that can store up to (expt 2 53) without loss of precision. You can't
use bit operations but you can add and multiply:

```
ELISP> (+ (* #xAABBCC (float (expt 2 24)))                                                                                                                               
          #x112233)
187723559281203.0
ELISP> (floor *)
187723559281203 (#o5253571404221063, #xaabbcc112233)
```
You can't store attributes in the same float too though, since you'll be exceeding range by 2 bits (you don't need to store the truecolor bit if you use a separate hash table). You could try a multi-level hash table. You should test only at load time if Emacs is 32bit or 64bit (e.g. (> most-positive-fixnum (expt 2 56.0))

Now this is starting to get very messy as you have two paths: 64bit Emacs with a single face cache, 32bit Emacs with two face caches and the xterm-color--with-ANSI-macro-helpers macro is already unwieldy. So in the interests of keeping the code simple, I say don't bother and don't add support for truecolor on 32bit Emacs. If you still want to give it a shot go ahead, but I won't merge it if I think that it overcomplicates  the rest of the code. It's simply not worth it.